### PR TITLE
Add `RasterBand::write_block`

### PR DESCRIPTION
**Pull request by lnicola**
- [x] I agree to follow the project's [code of conduct](https://github.com/georust/gdal/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

### CODE REVIEW

**Comment by lnicola** [comment]
Will rebase on top of #489.

r? @metasim

**Comment by metasim** [comment]
Should we also support `{read|write}_block` without `ndarray`?

File: src/raster/rasterband.rs [hunk 1]
```diff
@@ -556,6 +562,41 @@ impl<'a> RasterBand<'a> {
         Array2::from_shape_vec((size.1, size.0), data).map_err(Into::into)
     }
 
+    #[cfg(feature = "ndarray")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "array")))]
+    /// Write a [`Array2<T>`] from a [`Dataset`] block, where `T` implements [`GdalType`].
+    ///
+    /// # Arguments
```
Comment (metasim) [inline src/raster/rasterband.rs:569]: Can we add a reference to the method for getting the block layout (so one can know how many blocks there are in each dimension)?

File: src/raster/rasterband.rs [hunk 2]
```diff
@@ -556,6 +562,41 @@ impl<'a> RasterBand<'a> {
...
+    /// * `block_index` - the block index
+    ///
+    /// # Notes
+    /// The Matrix shape is (rows, cols) and raster shape is (cols in x-axis, rows in y-axis).
+    pub fn write_block<T: Copy + GdalType>(
+        &mut self,
+        block_index: (usize, usize),
```
Comment (metasim) [inline src/raster/rasterband.rs:576]: What happens if the block index is out of bounds? Do we get an error, or `panic` (or worse)?

File: src/raster/rasterband.rs (hunk 1 above)
Comment (metasim) [reply src/raster/rasterband.rs:569]: An example would be helpful here.

**Review by metasim — CHANGES_REQUESTED** [review]
I'm torn about having core operations that are only supported with in `ndarray` enabled, mode without a technical reason.

* Pros: easier to submit PRs.
* Cons: forces users to use ndarry when in reality

I'd like to use these features as a `Buffer` user.

But I admittedly have a bit of an axe to grind here. I would have preferred ndarray integration to have happened as a copy-free `Buffer::to_ndarray` method (for example), rather than having dual I/O methods for `Buffer` and for `Array2`.

File: src/raster/rasterband.rs (hunk 2 above)
Comment (lnicola) [reply src/raster/rasterband.rs:646]: There's no way it can panic, of course :-). Added a note in the docs.

File: src/raster/rasterband.rs (hunk 1 above)
Comment (lnicola) [reply src/raster/rasterband.rs:592]: Added an example and a note to `block_size` and `actual_block_size`.

**Review by metasim — APPROVED** [review]

**Comment by lnicola** [comment]
@metasim do you ever use the MD API, and would you expect it to require `ndarray`, or use some form of `Buffer3`, `BufferN` version?

**Comment by metasim** [comment]
> do you ever use the MD API, and would you expect it to require ndarray, or use some form of Buffer3, BufferN version?

@lnicola I do not. That's a good point (I didn't realize MD required it). For that API it seems reasonable to require `ndarray`, and would be silly to implement a bespoke multidimensional array capability just to avoid a dependency.

I wouldn't be adverse to `ndarray` being required for everything. But because we advertise it as an "optional" dependency, I don't think not using it should preclude access to basic functionality (i.e. block I/O).

**Comment by lnicola** [comment]
> I didn't realize MD required it

It doesn't, I should have checked before asking :pensive:.

**Comment by metasim** [comment]
PS: I think `ndarray` is fantastic. My position is more about it's optionality, and treating `Buffer` as a first-class citizen if face of that optionality.

**Pull request merged**